common/txmgr/finalizer.go

```diff
@@ -0,0 +1,235 @@
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) processHead(ctx context.Context, head types.Head[BLOCK_HASH]) error {
+	latestFinalizedHead := head.LatestFinalizedHead()
+	// Cannot determine finality without a finalized head for comparison
+	if latestFinalizedHead == nil || !latestFinalizedHead.IsValid() {
+		return fmt.Errorf("failed to find latest finalized head in chain")
+	}
+	earliestBlockNumInChain := latestFinalizedHead.EarliestHeadInChain().BlockNumber()
+	f.lggr.Debugw("processing latest finalized head", "block num", latestFinalizedHead.BlockNumber(), "block hash", latestFinalizedHead.BlockHash(), "earliest block num in chain", earliestBlockNumInChain)
+
+	// Retrieve all confirmed transactions, loaded with attempts and receipts
+	confirmedTxs, err := f.txStore.FindTransactionsByState(ctx, TxConfirmed, f.chainId)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve confirmed transactions: %w", err)
+	}
+
+	var finalizedTxs []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	// Group by block hash transactions whose receipts cannot be validated using the cached heads
+	receiptBlockHashToTx := make(map[BLOCK_HASH][]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE])
+	// Find transactions with receipt block nums older than the latest finalized block num and block hashes still in chain
+	for _, tx := range confirmedTxs {
+		// Only consider transactions not already marked as finalized
+		if tx.Finalized {
+			continue
+		}
+		receipt := tx.GetReceipt()
+		if receipt == nil || receipt.IsZero() || receipt.IsUnmined() {
+			continue
+		}
+		// Receipt newer than latest finalized head block num
+		if receipt.GetBlockNumber().Cmp(big.NewInt(latestFinalizedHead.BlockNumber())) > 0 {
+			continue
+		}
+		// Receipt block num older than earliest head in chain. Validate hash using RPC call later
+		if receipt.GetBlockNumber().Int64() < earliestBlockNumInChain {
+			receiptBlockHashToTx[receipt.GetBlockHash()] = append(receiptBlockHashToTx[receipt.GetBlockHash()], tx)
+			continue
+		}
+		blockHashInChain := latestFinalizedHead.HashAtHeight(receipt.GetBlockNumber().Int64())
+		// Receipt block hash does not match the block hash in chain. Transaction has been re-org'd out but DB state has not been updated yet
+		if blockHashInChain.String() != receipt.GetBlockHash().String() {
+			continue
+		}
+		finalizedTxs = append(finalizedTxs, tx)
+	}
+
+	// Check if block hashes exist for receipts on-chain older than the earliest cached head
+	// Transactions are grouped by their receipt block hash to minimize the number of RPC calls in case transactions were confirmed in the same block
+	// This check is only expected to be used in rare cases if there was an issue with the HeadTracker or if the node was down for significant time
+	var wg sync.WaitGroup
+	var txMu sync.RWMutex
+	for receiptBlockHash, txs := range receiptBlockHashToTx {
```

Comment:
Why not BatchCall instead?

Reply:
I would prefer using a batch call but I was trying to avoid introducing another method in the txmgr client just for this. Otherwise, using `BatchCallContext` directly would have introduced some chain specific code when specifying `eth_getBlockByHash`. Although, I'm reconsidering now if this component even needs to be in common. If we're pushing to degeneralize all of this code, I should put this in the EVM side where this isn't a problem.

Reply:
I moved the Finalizer to the EVM code and changed this to a batch call in the latest commit